src/rdkafka_zstd.c

```diff
@@ -76,23 +77,38 @@ rd_kafka_zstd_decompress (rd_kafka_broker_t *rkb,
                         return RD_KAFKA_RESP_ERR__CRIT_SYS_RESOURCE;
                 }
 
-
                 ret = ZSTD_decompress(decompressed, (size_t)out_bufsize,
                                       inbuf, inlen);
+
                 if (!ZSTD_isError(ret)) {
                         *outlenp = ret;
-                        *outbuf = decompressed;
+                        *outbuf  = decompressed;
+                        /* readjust memory usage because we might not have used the entire buffer */
+                        rd_restore_total_mem_usage(out_bufsize - ret);
+                        /* shrink buffer size to avoid memory waste */
+                        rd_realloc(decompressed, ret + 1);
```

Comment:
It could lead to fragmentation - I would not recommend doing it.

Reply:
OK, let me remove this